### PR TITLE
docs: repo guide & documentation

### PR DESCRIPTION
**Pull request by antoncoding**
## Moving Libraries:
* `AccountLib` is designed specifically to `CrossEngine`'s data structure, So i moved them to `engine/cross-margin` folder.
* since `Balance` is used by `Grappa`, it's now a shared structure among all engines, so I moved it to `BalanceUtil` in the `libraries` layer

## Docs
* View [here](https://github.com/antoncoding/grappa/tree/doc/repo-guide/docs)

### CODE REVIEW

File: docs/architecture.md [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+## Margin Engines
+
+**Margine Engines** are contracts that determine the rule to collateralize option tokens. Tokens minted by different engines are not fungible, so that the risk are always isolated. There should be multiple margin engines working together to provide user flexibilities to choose from, based on user preference such as gas fee, capital efficiency, composability and risk.
+
+### List of Margin Engines
+
+- `FullMargin`: fully collateralized margin. can be used to mint:
+  - covered call (collateralized with underlying)
+  - covered put (collateralized with strike)
```
Comment (dsshap) [inline docs/architecture.md:35]: is this `collateralized with strike` enforced?

File: docs/repo-guide.md [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+
+note: `subAccount`*: In the current shared engine design (inherited from `BaseEngine`), we use a `subAccount` address to map ID to a specific account data structure designed by each engine. A `subAccount` is "controllable" by an address if only the last 1 bytes are different from the address. This mean each address can control 256 subAccounts. This design is inspired by [Euler](https://github.com/euler-xyz/euler-contracts/blob/cd3036e0087280365819f99ad531141894d0b7ee/contracts/BaseLogic.sol#L24), it enables cheaper auth process without defining a `owner` for each ID.
+
+(You don't have to follow this access control design to be compatible with Grappa)
+
+## Integration tests
+
+You can find the integration tests for each engine in `test/engine-integrations/`. In these tests, we setup the engine with the real Grappa contract and OptionToken contract to test the margining rules.
+
+Mosts of the tests for engines are in this form, as the
```
Comment (dsshap) [inline docs/repo-guide.md:19]: ?

File: docs/architecture.md (hunk 1 above)
Comment (antoncoding) [reply docs/architecture.md:35]: yes

Comment (dsshap) [reply docs/architecture.md:35]: Where?

Comment (dsshap) [reply docs/architecture.md:35]: Oh I see it in your lib, the PUT check was getting cut off on my phone.

**Review by dsshap — APPROVED** [review]
LGTM!

I'll merge this in after my PR#160

**Pull request merged**